testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_concurrency.py

```diff
@@ -0,0 +1,209 @@
+import pytest
+
+from keywords.MobileRestClient import MobileRestClient
+from keywords import document
+from CBLClient.Replication import Replication
+from CBLClient.Authenticator import Authenticator
+from libraries.testkit import cluster
+from CBLClient.Document import Document
+
+
+def create_doc_from_document(doc_id, content, channel):
+    return document.create_doc(doc_id=doc_id, content=content, channels=channel)
+
+
+@pytest.mark.listener
+@pytest.mark.parametrize("concurrencyType", [
+    ("lastWriteWins"),
+    ("failOnConflict")
+])
+def test_replication_with_concurrencyControl_sameDocId_createUpdate(params_from_base_test_setup, concurrencyType):
+    """
+    @summary:
+    1. Create document id = doc1 as doc1a instance
+    2. Create document id = doc1 as doc1b instance
+    3. Save doc1a
+    4. Save doc1b (success for LastWriteWins, fail for FailOnConflict)
+    5. update document
+    """
+
+    cluster_config = params_from_base_test_setup["cluster_config"]
+    base_url = params_from_base_test_setup["base_url"]
+    sg_config = params_from_base_test_setup["sg_config"]
+    db = params_from_base_test_setup["db"]
+    cbl_db = params_from_base_test_setup["source_db"]
+
+    channel = ["Replication-1"]
+    doc_id = "doc_1"
+    documentObj = Document(base_url)
+
+    c = cluster.Cluster(config=cluster_config)
+    c.reset(sg_config_path=sg_config)
+
+    # 1. Create document id = doc1 as doc1a instance
+    doc_body = create_doc_from_document(doc_id=doc_id, content="doc1", channel=channel)
+    mutable_doc1 = documentObj.create(doc_id, doc_body)
+
+    # 2. Create document id = doc1 as doc1b instance
+    doc_body2 = create_doc_from_document(doc_id=doc_id, content="doc2", channel=channel)
+    mutable_doc2 = documentObj.create(doc_id, doc_body2)
+    db.saveDocumentWithConcurrency(cbl_db, mutable_doc1, concurrencyType)
+    db.saveDocumentWithConcurrency(cbl_db, mutable_doc2, concurrencyType)
+
+    # 3. Get cbl docs
+    cbl_doc_ids = db.getDocIds(cbl_db)
+    cbl_docs = db.getDocuments(cbl_db, cbl_doc_ids)
+    for id in cbl_doc_ids:
+        if concurrencyType == "lastWriteWins":
+            assert cbl_docs[id]["content"] == "doc2", "last write wins did not succeed"
+        else:
+            assert cbl_docs[id]["content"] == "doc1", "Fail on conflict did not work"
+
+    # 4. update doc with first update
+    document = db.getDocument(cbl_db, doc_id)
+    document2 = db.getDocument(cbl_db, doc_id)
```

Comment:
same doc as above, then why extra variable. You can pass document only unless you are modifying it somewhere.

Reply:
to test concurrency, you need to create two instances of same doc. document2 is used in following steps 72 -79. This is basically same instances taken by two threads and modifying

Reply:
ok with the explanation